architecture_records/002-acceleration-framework.md

```diff
@@ -0,0 +1,480 @@
+* See also [concrete plugin implementation that loads quantized model (using AutoGPTQ) for LoRA training](#detailed-design).
+* Even though all 3 methods are optional, at least one should be implemented.
+
+```python
+
+# data class to hold data and pointer to registered plugins
+@dataclass
+class PluginRegistration:
+    plugin: "AccelerationPlugin"
+    configuration_paths: List[str] # path 
+
+# global object to store all registered plugins
+PLUGIN_REGISTRATIONS: List[PluginRegistration] = list()
+
+# this is a base class from which concrete implementations will inherit from
+class AccelerationPlugin:
+
+    @staticmethod
+    def register_plugin(
+        plugin: "AccelerationPlugin", configuration_paths: List[str], 
+        **kwargs,
+    ):
+        global PLUGIN_REGISTRATIONS
+        PLUGIN_REGISTRATIONS.append(
+            PluginRegistration(plugin, configuration_paths)
+        )
+
+    # if specified, will restricted plugin to specified model archs
+    # - useful if method is restricted to certain model architectures, e.g., only used for MoEs
+    restricted_model_archs: Set = None
+
+    # if specified, will check if the package/s is/are installed
+    required_packages: Set = None
+
+    @property
+    def requires_custom_loading(self):
+        return False # to return True if plugin requires custom model 
+
+    @property
+    def requires_agumentation(self):
+        return False # to return True if plugin requires model augmentation
+
+    def model_loader(model_path: str, **kwargs):
+        pass # to be replaced with concrete
+
+    # augment model or accelerator object
+    def augmentation(model: nn.Module, **kwargs):
+        pass
+
+    def callbacks(model: nn.Module, **kwargs):
+        return []
+```
+
+
+#### Implmentation of Acceleration Framework Class
+
+The role of `AccelerationFramework` is to manage implemented plugins. In particular:
+- parse `configuration_file`, see below, and based on contents, decide the `AccelerationPlugin`'s to promote to `active_plugins`. See [implementaiton of AutoGPTQ LoRA Plugin](#detailed-design) for more description on configuration logic.
+- handle *plugin stacking*, i.e., when we have more than one `active_plugins`, apply their `model_loader` / `augmentation` logic in appropriate succession. 
+    * This is very useful, e.g., loading a LoRA-trainable AutoGPTQ model first, then applying additional optimized fused kernels to further improve training speeds.
+- enforce that `AccelerationFramework.model_loader` witll call `AccelerationPlugin.model_loader` of some `active_plugins` *at most once*.
+    * Prevents potential complication as multiple `active_plugins` could load models in conflicting ways.
+- enforce that `AccelerationFramework.augmentation` would apply `AccelerationPlugin.augmentation` in appropriate succession.
+    * c.f. previous example, where first a LoRA-trainable AutoGPTQ model is loaded, then fused_kernels on applied on top.
```

Comment:
so we dont want to limit number of augmentation techniques that can be applied? there can be multiple one after the other?

Reply:
correct, we want to have a way to compose plugins. One concrerte use case i have is
- have one plugin to load an autoGPTQ prepared model for lora
- have an option to additional load a plugin that installs the lora merge ops, loss kernels, etc, on top of the model loaded by first plugin

That way we can have multiple options, and see how the different plugins / acceleration impact the speed. Also it has implications on needs, if user wants it simple, and does not need the additional speed (because dont want to install additional cuda kernels), can opt out of the second plugin